File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode res = new ListNode();
        ListNode dummy = res;

        while(list1 != null && list2 != null) {
            if(list1.val > list2.val) {
                dummy.next = new ListNode(list2.val);
                list2 = list2.next;
            } else {
                dummy.next = new ListNode(list1.val);
                list1 = list1.next;
            }

            dummy = dummy.next;
        }

        ListNode remainder = list1 == null ? list2 : list1;

        dummy.next = remainder;

        return res.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        dummy = ListNode()
        cur = dummy

        for idx, l in enumerate(lists):
            if l:
                heapq.heappush(min_heap, (l.val, idx, l))
        
        while min_heap:
            val, i, node = heapq.heappop(min_heap)
            cur.next = node
            cur = cur.next
            if cur.next:
                heapq.heappush(min_heap, (cur.next.val, i, cur.next))
        
        return dummy.next
